jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java

```diff
@@ -1113,24 +1114,28 @@ else if (c == '/')
                         switch (c)
                         {
                             case ';':
-                                checkSegment(uri, segment, i, true);
+                                if (segment != i)
+                                    checkSegment(uri, segment, i, true);
```

Comment:
I don't think that is right.  If segment == i, that is an empty segment.  So I think now you are not checking empty segments.

Eitherway, I think that another state is better than having lots of replicated if statements

So maybe have 2 PATH states.  You need one state that any valid terminator is also the end of a segment (eg `/`, `?`, `#`, `;` and EOS) and then you have another state that can see EOS, `?` or `#` without it being the end of a segment. 

So if you had just finished parsing `/foo/bar/` you'd be in second of these states, so if the next character is `?` that is not an empty segment, but if the next character is `;` that is an empty segment.  IE `/foo/bar/?n=v` has 2 segments and `/foo/bar/;?n=v` has 3 segments with the last being empty.